File: internal/api/v1/capabilities.go
package v1

import (
	"net/http"

	"github.com/onscreen/onscreen/internal/api/respond"
)

// CapabilitiesResponse describes what this server can do. Public, intended
// for native clients that just discovered the server and need to know
// whether it's worth connecting to.
type CapabilitiesResponse struct {
	Server    CapabilitiesServer    `json:"server"`
	Features  CapabilitiesFeatures  `json:"features"`
	Codecs    CapabilitiesCodecs    `json:"codecs"`
	Limits    CapabilitiesLimits    `json:"limits"`
	Discovery CapabilitiesDiscovery `json:"discovery"`
}

// CapabilitiesServer carries server identity. Don't put anything sensitive
// here — the endpoint is anonymous.
type CapabilitiesServer struct {
	Name       string `json:"name"`
	MachineID  string `json:"machine_id"`
	Version    string `json:"version"`
	APIVersion string `json:"api_version"`
}

// CapabilitiesFeatures advertises the optional features clients can rely on.
// All bools so clients can do `if (caps.features.trickplay)` without having
// to interpret strings.
type CapabilitiesFeatures struct {
	Transcode         bool `json:"transcode"`
	Trickplay         bool `json:"trickplay"`
	SubtitlesExternal bool `json:"subtitles_external"`
	SubtitlesOCR      bool `json:"subtitles_ocr"`
	OIDC              bool `json:"oidc"`
	LDAP              bool `json:"ldap"`
	DevicePairing     bool `json:"device_pairing"`
	Plugins           bool `json:"plugins"`
	Backup            bool `json:"backup"`
	PeopleCredits     bool `json:"people_credits"`
	Photos            bool `json:"photos"`
	Music             bool `json:"music"`
	Webhooks          bool `json:"webhooks"`
	Notifications     bool `json:"notifications"`
	Requests          bool `json:"requests"`
	LiveTV            bool `json:"live_tv"`
	DVR               bool `json:"dvr"`
	Lyrics            bool `json:"lyrics"`
	IntroMarkers      bool `json:"intro_markers"`
	Chapters          bool `json:"chapters"`
}

// CapabilitiesCodecs advertises which codecs this server can transcode
// TO — clients match against these when picking a target profile. Listed
// explicitly rather than assumed so we can drop codecs at runtime based
// on detected encoders.
type CapabilitiesCodecs struct {
	Video         []string `json:"video"` // e.g. ["h264", "hevc"]
	Audio         []string `json:"audio"` // e.g. ["aac", "ac3", "mp3"]
	Containers    []string `json:"containers"`
	Hardware      []string `json:"hardware"` // active encoder names: h264_nvenc, h264_amf, ...
	HDRToneMap    bool     `json:"hdr_tonemap"`
}

// CapabilitiesLimits documents server-side caps so a client can pre-validate
// before asking the server to transcode at 8K and getting an error.
type CapabilitiesLimits struct {
	MaxUploadBytes          int64 `json:"max_upload_bytes"`
	MaxTranscodeBitrateKbps int   `json:"max_transcode_bitrate_kbps"`
	MaxTranscodeWidth       int   `json:"max_transcode_width"`
	MaxTranscodeHeight      int   `json:"max_transcode_height"`
	MaxConcurrentTranscodes int   `json:"max_concurrent_transcodes"`
	LiveTVTuneCount         int   `json:"live_tv_tune_count"`
}

// CapabilitiesDiscovery describes how to find this server again later
// without going through HTTP every time.
type CapabilitiesDiscovery struct {
	UDPPort int `json:"udp_port,omitempty"`
}

// CapabilitiesProvider is the small slice of config + state the handler
// needs. Injecting an interface (rather than the full Config) keeps the
// handler trivial to test.
type CapabilitiesProvider interface {
	Capabilities() CapabilitiesResponse
}

// CapabilitiesHandler serves GET /api/v1/system/capabilities.
type CapabilitiesHandler struct {
	provider CapabilitiesProvider
}

// NewCapabilitiesHandler constructs a handler.
func NewCapabilitiesHandler(p CapabilitiesProvider) *CapabilitiesHandler {
	return &CapabilitiesHandler{provider: p}
}

// Get handles the request.
func (h *CapabilitiesHandler) Get(w http.ResponseWriter, r *http.Request) {
	respond.Success(w, r, h.provider.Capabilities())
}

File: clients/android/app/src/main/java/tv/onscreen/android/data/model/Capabilities.kt
package tv.onscreen.android.data.model

import com.squareup.moshi.JsonClass

/**
 * Server capabilities — feature flags, supported codecs, and operational
 * limits. Returned by GET /api/v1/system/capabilities.
 *
 * The server promises forward-compatibility on this shape: v2.x minor
 * releases may add fields but won't remove or rename them. Moshi's
 * loose decoding (default values per field) plus this contract means
 * a TV running an older binary against a newer server still gets a
 * usable subset.
 *
 * Per-flag rationale lives on the server's openapi.yaml — the
 * feature flags that AND-gate against runtime tool probes (`trickplay`,
 * `subtitles_ocr`, `intro_markers`, `backup`) reflect what *actually*
 * works on the box, not what the codebase intends to support.
 */
@JsonClass(generateAdapter = true)
data class Capabilities(
    val server: ServerInfo = ServerInfo(),
    val features: CapabilitiesFeatures = CapabilitiesFeatures(),
    val codecs: CapabilitiesCodecs = CapabilitiesCodecs(),
    val limits: CapabilitiesLimits = CapabilitiesLimits(),
    val discovery: CapabilitiesDiscovery = CapabilitiesDiscovery(),
)

@JsonClass(generateAdapter = true)
data class ServerInfo(
    val name: String = "",
    val machine_id: String = "",
    val version: String = "",
    val api_version: String = "",
)

@JsonClass(generateAdapter = true)
data class CapabilitiesFeatures(
    val transcode: Boolean = false,
    val trickplay: Boolean = false,
    val subtitles_external: Boolean = false,
    val subtitles_ocr: Boolean = false,
    val oidc: Boolean = false,
    val ldap: Boolean = false,
    val device_pairing: Boolean = false,
    val plugins: Boolean = false,
    val backup: Boolean = false,
    val people_credits: Boolean = false,
    val photos: Boolean = false,
    val music: Boolean = false,
    val webhooks: Boolean = false,
    val notifications: Boolean = false,
    val requests: Boolean = false,
    val live_tv: Boolean = false,
    val dvr: Boolean = false,
    val lyrics: Boolean = false,
    val intro_markers: Boolean = false,
    val chapters: Boolean = false,
    // web_downloads is server-side admin posture for the watch-page
    // Download button. TV clients don't surface it (no save-as on a
    // stationary set-top device) but the field is decoded so it
    // doesn't trip the strictness setting.
    val web_downloads: Boolean = false,
)

@JsonClass(generateAdapter = true)
data class CapabilitiesCodecs(
    val video: List<String> = emptyList(),
    val audio: List<String> = emptyList(),
    val containers: List<String> = emptyList(),
    val hardware: List<String> = emptyList(),
    val hdr_tonemap: Boolean = false,
)

@JsonClass(generateAdapter = true)
data class CapabilitiesLimits(
    val max_upload_bytes: Long = 0L,
    val max_transcode_bitrate_kbps: Int = 0,
    val max_transcode_width: Int = 0,
    val max_transcode_height: Int = 0,
    val max_concurrent_transcodes: Int = 0,
    val live_tv_tune_count: Int = 0,
)

@JsonClass(generateAdapter = true)
data class CapabilitiesDiscovery(
    val udp_port: Int = 0,
)
